Integrate sample metadata API

vivbak opened the pull request:
Integrating Sample Metadata API. Re-writing existing scripts for this implementation. 

- [x] Re-write [Upload Processor](https://github.com/populationgenomics/joint-calling/blob/main/joint_calling/upload_processor.py) 
- [x] Re-write [Process Uploads](https://github.com/populationgenomics/joint-calling/blob/main/scripts/process_uploads.py)
 ~~Create Upload Samples script to handle csv separately.~~ [Replaced by generic ingester]
- [x] Update broken test file. 
~~Modify [find inputs]( https://github.com/populationgenomics/joint-calling/blob/301d4d05f9570a671e5f98aa4ed1e776e7b549da/joint_calling/utils.py#L81 )~~  [Moved to different PR]

vivbak commented:
Hiya @illusional :) 

Could you take a quick peek and let me know if these are the sorts of changes we were looking for? 
I've commented `#TODO: Implement.` every time we use an endpoint that needs to be built. 

Let me know if this is what we were aiming for. If so, I can finish up the changes for the other scripts.

Thanks!

illusional commented on joint_calling/upload_processor.py:
```diff
@@ -69,9 +73,14 @@ def batch_move_files(
-        j = batch.new_job(name=f'move {sample}')
+        # Get internal sample ID
+        sapi = SampleApi()
+        internal_id = sapi.get_internal_id(sample)  # TO IMPLEMENT: API CALL
```
It might be quicker to batch these calls at the start, eg:

```python
internal_id_map = sapi.get_internal_ids_from_external_ids(samples)
for external_sample_id in samples:
    internal_id = internal_id_map[external_sample_id]
```

vivbak replied:
Great point! Pulled it out :)

illusional commented on scripts/process_uploads.py:
```diff
@@ -3,89 +3,63 @@
 
+    # Returns a list of all samples with status == uploaded
+    # Alternatively, could return all samples that were not included in the latest .mt analysis.
+    samples = sapi.get_latest_uploads()  # TO IMPLEMENT: API CALL
```
We want to be a bit clearer here, a sample doesn't have a status (because it's a physical status). The status = uploaded would be a byproduct of either the gvcf or cram being uploaded, so this call might look like:

- `get_samples_which_have_gvcfs(project)`

illusional commented on scripts/process_uploads.py:
```diff
@@ -3,89 +3,63 @@
 
 def generate_file_list(
-    samples: Set[str],
+    samples: List[str],
```
It would be worth clarifying this variable, eg: `internal_sample_ids` (and the subsequent `s` -> `internal_sample_id`)

vivbak replied:
Have renamed the variables in this function to hopefully add clarity :D

illusional commented on scripts/upload_samples.py:
```diff
@@ -0,0 +1,95 @@
+
+    for sample in latest_upload:
+        # pylint: disable=W0640
+        current_metadata = filter(
```
Instead of filtering the array of `sample_metadata` to find the ID for each time, consider using a map, to first key by the sample_name, and then you can pull it out in one hit, eg:

```python
sample_meta_map = {d['sample.sample_name']: d for d in sample_metadata}
for sample in latest_upload:
    metadata = sample_meta_map[sample]
    # ...
```

vivbak replied:
That's much nicer! Thank you :)

vivbak commented on joint_calling/upload_processor.py:
```diff
@@ -69,9 +73,14 @@ def batch_move_files(
     jobs = []
     for sample in files:
```
Great spot! This was from when we previously just iterated through a list of samples. I've modified the implementation slightly. The equivalent variable is now called `file_name` and it's an external sample id + file extension.

vivbak commented on scripts/upload_samples.py:
```diff
@@ -0,0 +1,95 @@
+    # deleted_samples = mt_samples - samples
+
+    # Determine the samples in the latest upload.
+    latest_upload = samples - mt_samples
```
Yeah, I think I've had this mix-up in a few places -- have tried renaming things so that there are less jumbles :)

vivbak commented:
Hi @illusional :D Huge thanks for your review and your patience with me addressing the feedback! I believe most things have now been properly addressed. However, there is one point that you raised a few times that so I just wanted to explain my implementation and get your feedback: 

`upload_samples.py` will ingest the CSV file, and update SM DB to let it know which analysis files (i.e. gvcfs and crams) are ready to be processed. 

In the latest revision it:
1. Pulls a list of internal sample IDs that don't have gvcfs 
2. Maps the list above to their external IDs
3. Based on the list identified in 2, determines the samples that are now ready to be processed. 
4. Maps back to the internal id's, creates a gvcf and cram analysis object for each of these samples. 
5. Updates the sequencing status to 'Ready' for each SampleSequence. 

`process_uploads.py `will generate the batch jobs required to perform the move for the samples identified in `upload_samples.py`
In the latest revision it: 
1. Determines the samples that are ready to be processed by pulling all of the SampleSequence objects with Status 'Ready' as defined in step 5 above. Assume this returns a series of internal sample IDs. 
2. Maps the IDs obtained above to their external IDs. 
3. Runs the rest of the processor. 
4. When the files are successfully processed the sequencing status is updated to 'Uploaded'. 

One thing that is still not clear in my mind, is that we're updating and reading the SampleSequence status in this approach. However, we are providing a SampleID.  Given that there can be multiple SampleSequences for a given sample, I feel like that doesn't work but I'm not super sure of the best approach to pull that SampleSequenceID.

Step 4 in `process_uploads.py` is also problematic because moving the gvcf files and moving the cram files are separated into different jobs that are independent of each other -- so knowing when to update this status is still a little unclear to me.

illusional commented on joint_calling/upload_processor.py:
```diff
@@ -67,11 +72,23 @@ def batch_move_files(
+    internal_id = sample_group.sample_id_internal
+    batch_number = sample_group.batch_number
+
+    for tuple_key in sample_group._fields:
+        if tuple_key in ['sample_id_external', 'batch_number', 'sample_id_internal']:
+            continue
+        file_name = getattr(sample_group, tuple_key)
```
It'd be clearer to define what files are being moved, rather than having to refer back to the sample_group definition to determine.

```suggestion
    files_to_move = (sample_group.data_file, sample_group.index_file, sample_group.md5)
    for file_name in files_to_move:
```

illusional reviewed (changes requested):
Looking great overall!

I've suggested a few naming changes, I believe it would be clearer to use named arguments when instantiating a SampleGroup, but that's my opinion and not a rule, eg:

```py
SampleGroup(
    sample_id_external=external_id,
    sample_id_internal=internal_id,
    data_file=f'{external_id}.g.vcf.gz',
    index_file=f'{external_id}.g.vcf.gz.tbi',
    md5=f'{external_id}.g.vcf.gz.md5',
    batch_number=batch_id,
)
```

I think the invocations ~~upload_analysis~~ `create_analysis_in_sm_db` is missing the `proj` parameter.

vivbak commented:
Hey @illusional, thanks for taking the time to review! I've gone ahead and applied most of the changes. One note is the test file actually needs to be reworked completely. I'll do that soon.

illusional commented on scripts/process_uploads.py:
```diff
@@ -35,28 +37,26 @@ def determine_samples(proj):
-
-        except ServiceException:
-            print(f'Sequencing for {sample} not found')
+    for internal_sample_id in samples_without_analysis['sample_ids']:
+        seq_id = int(seqapi.get_sequence_id_from_sample_id(internal_sample_id, proj))
+        seq_entry = seqapi.get_sequence_by_id(seq_id, proj)
```
I'm not really a fan of lots of web requests in the loop, I've just added `get_sequences_by_internal_sample_ids` (will require regeneration of SM) unfortunately to hopefully dramatically reduce the amount of web requests.

```suggestion
    sequences = seqapi.get_sequences_by_internal_sample_ids(
        project=proj, request_body=samples_without_analysis['sample_ids']
    )
    for seq_entry in sequences:
        internal_sample_id = seq_entry.sample_id
```

illusional commented on scripts/process_uploads.py:
```diff
@@ -99,21 +99,21 @@ def generate_file_list(
-            f'{external_id}.g.vcf.gz.tbi',
-            f'{external_id}.g.vcf.gz.md5',
-            batch_id,
+            sample_id_external=external_id,
```
🙏

illusional commented on scripts/process_uploads.py:
```diff
@@ -35,28 +37,26 @@ def determine_samples(proj):
+            if 'gvcf' in seq_entry['meta']:
+                batch = seq_entry['meta']['gvcf'].get('batch')
+                samples_with_sequencing_meta.append(internal_sample_id)
+                external_map = sapi.get_sample_id_map_by_internal(
```
Can this request be done outside the loop, with all sample ids?

illusional reviewed (changes requested):
Create file list from sequence.meta payload as discussed.

illusional commented on joint_calling/upload_processor.py:
```diff
@@ -67,11 +78,27 @@ def batch_move_files(
+        # previous_location = os.path.join('gs://', source_prefix, file_name)
+        # previous_location = file_name
+
+        file_extension = base[len(external_id) :]
```
Make sure you import os.path at the top of the file.

```suggestion
        file_extension = os.path.splitext(base)[1]
```

illusional commented on scripts/process_uploads.py:
```diff
@@ -3,95 +3,151 @@
+        file_extension = '.cram'
 
-    return job
+    filepath = os.path.join('gs://', path, internal_id + file_extension)
```
Can you pass the fully formed gs path to the file into this method, as "output_path" instead of constructing it inside, and outside. If you change the format in one place, it would be difficult to remember to change the other.

illusional reviewed (changes requested):
Looks great, I've asked for two small changes:
- Better extension detection
- Constructing the fully qualified path before creating the analysis object.

illusional reviewed (approved):
Awesome work!!
